solr/core/src/java/org/apache/solr/response/transform/DeeplyNestedChildDocTransformerFactory.java

```diff
@@ -0,0 +1,367 @@
+      } catch (SyntaxError syntaxError) {
+        throw new SolrException( ErrorCode.BAD_REQUEST, "Failed to create correct child filter query" );
+      }
+    }
+
+    String parentFilter = params.get( "parentFilter" );
+
+    BitSetProducer parentsFilter = null;
+
+    if(parentFilter != null) {
+      try {
+        Query parentFilterQuery = QParser.getParser( parentFilter, req).getQuery();
+        //TODO shouldn't we try to use the Solr filter cache, and then ideally implement
+        //  BitSetProducer over that?
+        // DocSet parentDocSet = req.getSearcher().getDocSet(parentFilterQuery);
+        // then return BitSetProducer with custom BitSet impl accessing the docSet
+        parentsFilter = new QueryBitSetProducer(parentFilterQuery);
+      } catch (SyntaxError syntaxError) {
+        throw new SolrException( ErrorCode.BAD_REQUEST, "Failed to create correct parent filter query" );
+      }
+    } else {
+      String sRootFilter = "{!frange l=1 u=1}strdist(" + req.getSchema().getUniqueKeyField().getName() + "," + ROOT_FIELD_NAME + ",edit)";
+      try {
+        Query parentFilterQuery = QParser.getParser(sRootFilter, req).getQuery();
+        //TODO shouldn't we try to use the Solr filter cache, and then ideally implement
+        //  BitSetProducer over that?
+        // DocSet parentDocSet = req.getSearcher().getDocSet(parentFilterQuery);
+        // then return BitSetProducer with custom BitSet impl accessing the docSet
+        parentsFilter = new QueryBitSetProducer(parentFilterQuery);
+      } catch (SyntaxError syntaxError) {
+        throw new SolrException( SolrException.ErrorCode.BAD_REQUEST, "Failed to create correct parent filter query" );
+      }
+    }
+
+    if(childFilterQuery == null) {
+      return new DeeplyNestedChildDocTransformer(field, parentsFilter, req, limit);
+    }
+    return new DeeplyNestedFilterChildDocTransformer(field, parentsFilter, req, childFilterQuery, nestPath!=null? generatePattern(splitPath): null, limit);
+  }
+
+  private Pattern generatePattern(List<String> pathList) {
+    if(pathList.size() <= 2) {
+      return Pattern.compile(pathList.get(0) + NUM_SEP_CHAR + "\\d");
+    }
+    return Pattern.compile(String.join(NUM_SEP_CHAR + "\\d" + PATH_SEP_CHAR, pathList.subList(0, pathList.size() - 1)) + NUM_SEP_CHAR + "\\d");
+  }
+}
+
+class DeeplyNestedFilterChildDocTransformer extends DeeplyNestedChildDocTransformerBase {
+
+  private Query childFilterQuery;
+  private Pattern nestPathMatcher;
+
+  public DeeplyNestedFilterChildDocTransformer( String name, final BitSetProducer parentsFilter,
+                              final SolrQueryRequest req, final Query childFilterQuery, Pattern pathPattern, int limit) {
+    super(name, parentsFilter, req, limit);
+    this.childFilterQuery = childFilterQuery;
+    this.nestPathMatcher = pathPattern;
+  }
+
+  @Override
+  public void transform(SolrDocument rootDoc, int docid) {
```

Comment:
This is a going to be a real challenge to come up with something that is understandable (doesn't appear *too* complicated).  Right now it appears too complicated to me; it's hard to review.  Tonight I think I'll propose some pseudocode on how to structure it.

Reply:
Sure thing, any help would be welcome.
This is a pretty rough draft, so I will try to work on this ASAP before I add more logic to the transformer.